File: docs/static/examples/src/fireworks/ts/firework.ts
import { noStroke, random } from 'p8g.js';
import Victor from 'victor';
import Particle from './particle';

export default class Firework {
  private readonly _rocket: Particle;
  private _exploded: boolean;
  private readonly _particles: Particle[];

  constructor(rocket: Particle) {
    this._rocket = rocket;
    this._exploded = false;
    this._particles = [];
  }

  applyForce(force: Victor): void {
    if (!this._exploded) {
      this._rocket.applyForce(force);
    } else {
      for (const particle of this._particles) {
        particle.applyForce(force);
      }
    }
  }

  update(): void {
    if (!this._exploded) {
      this._rocket.update();
      if (this._rocket.velocity.y >= 0) {
        this.explode();
      }
    } else {
      for (const particle of this._particles) {
        particle.velocity.multiply(new Victor(0.95, 0.95));
        particle.update();
      }
      let index = -1;
      while (
        (index = this._particles.findIndex((particle) => particle.isDead())) !==
        -1
      ) {
        this._particles.splice(index, 1);
      }
    }
  }

  explode(): void {
    const color: [number, number, number, number] = [random(255), random(255), random(255), 255];
    for (let i = 0; i < 200; i++) {
      const angle = random(2 * Math.PI);
      const magnitude = random(7);
      const particle = new Particle(
        this._rocket.position.clone(),
        random(0.7, 1),
        new Victor(magnitude * Math.cos(angle), magnitude * Math.sin(angle)),
        Math.floor(random(90, 110)),
        color,
      );
      this._particles.push(particle);
    }
    this._exploded = true;
  }

  draw(): void {
    noStroke();
    if (!this._exploded) {
      this._rocket.draw();
    } else {
      for (const particle of this._particles) {
        particle.draw();
      }
    }
  }

  isDone(): boolean {
    return this._exploded && this._particles.length === 0;
  }
}

File: docs/static/examples/src/fireworks/java/Firework.java
import org.joml.Vector2f;
import org.joml.Vector4f;

import java.util.HashSet;
import java.util.Set;

import static io.github.bernhardfritz.p8g.Sketch.*;

public class Firework {

    private Particle rocket;
    boolean exploded;
    private Set<Particle> particles;

    public Firework(Particle rocket) {
        this.rocket = rocket;
        this.exploded = false;
        this.particles = new HashSet<>();
    }

    public void applyForce(Vector2f force) {
        if (!exploded) {
            rocket.applyForce(force);
        } else {
            for (Particle particle : particles) {
                particle.applyForce(force);
            }
        }
    }

    public void update() {
        if (!exploded) {
            rocket.update();
            if (rocket.getVelocity().get(1) >= 0.f) {
                explode();
            }
        } else {
            for (Particle particle : particles) {
                particle.getVelocity().mul(0.95f);
                particle.update();
            }
            particles.removeIf(Particle::isDead);
        }
    }

    public void explode() {
        Vector4f color = new Vector4f(random(255.f), random(255.f), random(255.f), 255.f);
        for (int i = 0; i < 200; i++) {
            float angle = random(2.f * (float) Math.PI);
            float magnitude = random(7.f);
            Particle particle = new Particle(
                    new Vector2f(rocket.getPosition()),
                    random(0.7f, 1.f),
                    new Vector2f(magnitude * (float) Math.cos(angle), magnitude * (float) Math.sin(angle)),
                    (int) random(90.f, 110.f),
                    color
            );
            particles.add(particle);
        }
        exploded = true;
    }

    public void draw() {
        noStroke();
        if (!exploded) {
            rocket.draw();
        } else {
            for (Particle particle : particles) {
                particle.draw();
            }
        }
    }

    public boolean isDone() {
        return exploded && particles.isEmpty();
    }

}
